### PR TITLE
Fix internal link issue

### PR DESCRIPTION
**Pull request by gjtorikian**
_Finally_ closes https://github.com/gjtorikian/html-proofer/issues/702.

In the end, this wasn't a blocker for 4.0, as I didn't need to change the cache format at all. Still, it was good to fix this one. Essentially, when it comes to internal links, we need to pay a _bit_ more attention as to the current_path, and not assume that an existing internal link is good. In other words, File A and File C can both point to File B in the same way (`a href="filec.html`), but we need to be sure that the paths of File A and File C can actually match what metadata is stored in the cache. 

@riccardoporreca if I could trouble you for a confirmation on this branch that would be great. 🙏

### CODE REVIEW

**Comment by riccardoporreca** [comment]
@gjtorikian, sure, I will look at this in the coming days (within this week), it deserves a bit of extra thoughts since managing the "found" information within the array of cached metadata can be still quite delicate

**Review by riccardoporreca — CHANGES_REQUESTED** [review]
@gjtorikian, I had a look at this, and did some concrete suggestions mainly in the context of ensuring the consistency of metadata with what the currently-detected URLs are. This is important for error logging in general bu more specififcally for the internal cache where the metadata are used to actually store "found" information depeding on the source file context.

The approach seems reasonable overall, although it still suffers conceptually from using as cache keys the href, which does not uniquely define the things that matter for internal links, implying we actually check all individual internal URLs with the same href (even e.g. the same internal link from the same page). My original proposal was trying to address this but was of course making things more complicated on the cache key definition, although simplifying the rest and making the logic for caching and re-checking aligned with the external links.

The main open issue I see is that we currently would still re-check all internal links having the same key (href) regardless of whether they were cached previously, provided at least one _needs_ being re-checked (due to being a new link or not being found previously). If relevant (perhaps it does not matter much for internal links), think might be solved by checking whether "found" is `true` before doing the check in `run_internal_link_checker`, as `true` would come the cache only with the new logic

File: lib/html_proofer/cache.rb [hunk 1]
```diff
@@ -157,6 +143,43 @@ def size(type)
...
+
+          # cache for this url, from an existing path, exists as found
+          if !existing_cache_metadata.nil? && !existing_cache_metadata.empty? && existing_cache_metadata[:found]
+            metadata.find { |m| m[:current_path] == existing_cache_metadata[:current_path] }[:found] = true
+            next
+          end
+
+          @logger.log(:debug, "Adding #{incoming_url} to internal cache")
+          arr << incoming_url
+        end
```
[the remark this answers is not in the document]
Comment (riccardoporreca) [reply lib/html_proofer/cache.rb:177]: Another thing: would it make sense to align the name of the "filename" and "current_path" keys in the metadata of internal and external URLs?
https://github.com/gjtorikian/html-proofer/blob/2873f06e1ddd02b09274ad8825becc6540a688c1/lib/html_proofer/check.rb#L54-L75

**Comment by gjtorikian** [comment]
If I am not mistaken, most of the concerns here revolve around performance, which, while important, is less important than all of the other changes I have for the 4.0 release. Given that the original bug was fixed, I'm going to merge this and work on other perf related tasks as time allows.

(I may, of course, be utterly wrong about all of that, so I'm open to pushback!)

**Comment by riccardoporreca** [comment]
Great to see the new release :rocket:, I am happy to upgrade on a few projects and monitor if there is any gap / misbehavior beyond performance.

**Pull request merged**